portal/flatpak-portal.c

```diff
@@ -999,7 +1029,28 @@ handle_spawn (PortalFlatpak         *object,
       env = g_strdupv (empty);
     }
   else
-    env = g_get_environ ();
+    {
+      GStrv run_environ;
+
+      if (testing)
+        {
+          static char *mock_run_environ[] = { "FOO=bar", NULL };
+          run_environ = mock_run_environ;
+        }
+      else
+        {
+          run_environ = flatpak_instance_get_run_environ (instance);
+          if (!run_environ)
+            {
+              g_dbus_method_invocation_return_error (invocation, G_DBUS_ERROR,
+                                                     G_DBUS_ERROR_INVALID_ARGS,
+                                                     "Could not find run environment of caller instance");
+              return G_DBUS_METHOD_INVOCATION_HANDLED;
+            }
+        }
+
+      env = g_strdupv (run_environ);
```

Comment:
This will need to become `g_strdupv ((GStrv) run_environ)` if we change `flatpak_instance_get_run_environ` to return `const char * const *`, but that's OK.